packages/expect/src/index.js

```diff
@@ -212,10 +213,46 @@ const makeThrowingMatcher = (
         utils,
       },
     );
-    let result: ExpectationResult;
+
+    const processResult = (result: SyncExpectationResult) => {
+      _validateResult(result);
+
+      getState().assertionCalls++;
+
+      if ((result.pass && isNot) || (!result.pass && !isNot)) {
+        // XOR
+        const message = getMessage(result.message);
+        const error = new JestAssertionError(message);
+        // Passing the result of the matcher with the error so that a custom
+        // reporter could access the actual and expected objects of the result
+        // for example in order to display a custom visual diff
+        error.matcherResult = result;
+        // Try to remove this function from the stack trace frame.
+        // Guard for some environments (browsers) that do not support this feature.
+        if (Error.captureStackTrace) {
+          Error.captureStackTrace(error, throwingMatcher);
+        }
+
+        if (throws) {
+          throw error;
+        } else {
+          getState().suppressedErrors.push(error);
+        }
+      }
+    };
+
+    let potentialResult: ExpectationResult;
 
     try {
-      result = matcher.apply(matcherContext, [actual].concat(args));
+      potentialResult = matcher.apply(matcherContext, [actual].concat(args));
+
+      if (isPromise(potentialResult)) {
+        potentialResult.then(aResult => {
```

Comment:
we need to `await` this somehow, and should probably have a `catch` as well

Reply:
Agree!